doc/modules/calibration.rst

```diff
@@ -62,7 +76,7 @@ by showing the number of samples in each predicted probability bin.
 
 :class:`LogisticRegression` returns well calibrated predictions by default as it has a
 canonical link function for its loss, i.e. the logit-link for the :ref:`log_loss`.
```

Comment:
I think we should be more precise here. This property only holds asymptotically. It could very well be the case when n_features is large compared to n_samples (overconfidence based and/or when the regularization is not well tuned.

In this specific plot, we can observe that the calibration of LogisticRegression is still far from perfect (it's overconfident), probably because the training set is very small (100 points and 20 features, with some noisy features) and that we use the default regularization without any tuning.

> Assuming that the number of training points is large enough and that the regularization is well tuned on a proper scoring rule such as the loglikelihood or the Brier score, LogisticRegression returns well calibrated predictions by default...

or more simply:

> Asymptotically (when the number of points in the training set size gets large enough), LogisticRegression returns well calibrated predictions by default... 

and do not mention regularization to simplify the message here.

Anyways, we probably need an extension to one of the examples about classifier calibration to study the joint impact of regularization tuning on proper scoring rules and the reliability diagrams to contrast over-regularized, under-regularized and well tuned LogReg models. This can be done in a follow-up PR.

Reply:
The point is that the balance property is not an asymptotic property. It holds exactly on the training data and no penalty. That means that every feature (at least the categorical once) of the design matrix is well conditionally calibrated. Note that conditional calibration implies auto-calibration which is assessed by the reliability diagram.

The question is then how good the training data represents the application data.

> In this specific plot, we can observe that the calibration of LogisticRegression is still far from perfect (it's overconfident)

For small (test) datasets, we should provide confidence intervals as done in https://lorentzenchr.github.io/model-diagnostics/examples/regression_on_workers_compensation/#31-reliability-diagrams.

Reply:
Interesting, let's keep this discussion for a follow-up PR then. I am pretty sure that tuning the regularization can have an impact on the shape of the reliability curve and that one cannot expect Log Reg to be always optimal when n_samples < n_features and noise. In that case I expect Log Reg to be overconfident.

> It holds exactly on the training data and no penalty. 

Users don't care much about the calibration of a predictive model on the training data. We want a well calibrated model on held-out data.